### ISSUE TITLE
No error is reported in local code base when destructing undefined while using downlevelIteration, but an error is reported in TypeScript Playground with same tsconfig

### ISSUE THREAD

**Issue by mrttrifork** [microsoft/TypeScript#34956]
**TypeScript Version:**  3.7.2 and 3.8.0-dev.20191105

**Search Terms:**
downlevelIterator, array destruction

**Code**

```ts
interface Dog {
    puppies: string[];
}

interface IAnimal {
    getDog(): Dog | null | undefined;
}

class Animal implements IAnimal {
    public getDog(): Dog | null | undefined {
        return undefined;
    }
}

class AnimalContainer {
    public readonly animal: Animal;

    constructor() {
        this.animal = new Animal();
        const [firstPup] = this.animal.getDog()?.puppies;
    }
}

const animalContainer = new AnimalContainer();

const [firstPuppy] = animalContainer.animal.getDog()?.puppies;
```

**Expected behavior:**

In the local code base this should have been the error message, as seen on TS Playground

Type 'string[] | undefined' is not an array type or does not have a '[Symbol.iterator]()' method that returns an iterator.

**Actual behavior:**

No error is emitted unless i set `downlevelIteration: false`

Then i get 

TS2569: string[] is not an array type or string type

**Playground Link:** 

https://www.typescriptlang.org/play/?noImplicitAny=false&strictFunctionTypes=false&strictPropertyInitialization=false&strictBindCallApply=false&noImplicitThis=false&noImplicitReturns=false&alwaysStrict=false&downlevelIteration=true&experimentalDecorators=true&emitDecoratorMetadata=true&target=1&module=1&ssl=1&ssc=1&pln=26&pc=63#code/JYOwLgpgTgZghgYwgAgCIHsDmyDeBYAKGWOQAcBXU04CAZwC40sBtAXQG5CBfQw0SWIhQBJAIIhgAWzgAbXIRLJMEMBkwAKAJSM1yAD7IQ5GXIPkQAEwgxQEC5wI8ChBDLi1aycVNnIppGQhJCHBPMQlpGXwiEgoAIxlgBCUVNS0dLH1DY1NkcysbEDt5GMViKBVyKBA8y2tbewUSJycXNw8vCNkAYXRwOFsoaMV4xOSKuAs+mQBPZDgumUZvSIcm4gQ+2jAocgQwdCgtErKSMAALYFoAOgWfOQBeQwgAd077rQdTja2wZGYbFBtgAFSisZBPC5XW6La7KVRYLQAfmuFCoNFoX2a3F4BE2IG280WvX6gwhzzeKx6fTAAyKR00azxv3+gJBlFIM3BTzukRJtMGMPucNSiM0KLR1Do7CAA

**Related Issues:** 

I couldn't find a similar issue after searching for array destruct and downlevelIterator

**Comment by RyanCavanaugh** [comment]
@rbuckton this regressed (!)
```ts
// @strict: true

declare const arr: string[] | undefined;
// No error, but should
const [x] = arr;
// Error
const x2 = arr[0];
```

**Comment by rbuckton** [comment]
```ts repro
// @strict: true
declare const arr: string[] | undefined;
// @ts-expect-error
const [x] = arr;
// @ts-expect-error
const x2 = arr[0];
```

[Workbench Repro](https://www.typescriptlang.org/dev/bug-workbench/?#code/PTAEAEGcBcCcEsDG0Bco4FcCmBYAUACZaIA2AhrFqIgPYB2MoFsaMCdA5gNoC6oAPqAx0iAM3h0sBANz4QEaJAC0WAB4AHYtBWxYNWPloNooLqr4BeJrtl554RSo1adeg3iONVAJlBXmXAAMPNJAA)

**Comment by rbuckton** [comment]
This seems to have been addressed in 4.1, as an error is now reported:
> ![image](https://user-images.githubusercontent.com/3902892/102291971-1977ea00-3ef9-11eb-8fdf-495eec3dd1c9.png)

**Comment by rbuckton** [comment]
It looks like this was resolved by #40350.
